x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java

```diff
@@ -173,69 +175,123 @@ private AuthorizationInfo getAuthorizationInfoFromContext() {
-                    assert false : message;
-                    listener.onFailure(new ElasticsearchSecurityException(message));
-                    return;
-                }
+            final String auditId;
+            try {
+                auditId = requireAuditId(authentication, action, originalRequest);
+            } catch (ElasticsearchSecurityException e) {
+                listener.onFailure(e);
+                return;
             }
 
             // sometimes a request might be wrapped within another, which is the case for proxied
             // requests and concrete shard requests
             final TransportRequest unwrappedRequest = maybeUnwrapRequest(authentication, originalRequest, action, auditId);
 
-            // Check operator privileges
-            // TODO: audit?
-            final ElasticsearchSecurityException operatorException =
-                operatorPrivilegesService.check(action, originalRequest, threadContext);
-            if (operatorException != null) {
-                listener.onFailure(denialException(authentication, action, originalRequest, operatorException));
+            if (checkOperatorPrivileges(authentication, action, originalRequest, listener) == false) {
                 return;
             }
-            operatorPrivilegesService.maybeInterceptRequest(threadContext, originalRequest);
 
             if (SystemUser.is(authentication.getUser())) {
                 // this never goes async so no need to wrap the listener
                 authorizeSystemUser(authentication, action, auditId, unwrappedRequest, listener);
             } else {
-                final String finalAuditId = auditId;
-                final RequestInfo requestInfo = new RequestInfo(authentication, unwrappedRequest, action);
+                final RequestInfo requestInfo = new RequestInfo(authentication, unwrappedRequest, action, parentContext);
+                final AuthorizationEngine engine = getAuthorizationEngine(authentication);
                 final ActionListener<AuthorizationInfo> authzInfoListener = wrapPreservingContext(ActionListener.wrap(
                         authorizationInfo -> {
                             threadContext.putTransient(AUTHORIZATION_INFO_KEY, authorizationInfo);
-                            maybeAuthorizeRunAs(requestInfo, finalAuditId, authorizationInfo, listener);
+                            maybeAuthorizeRunAs(requestInfo, auditId, authorizationInfo, listener);
                         }, listener::onFailure), threadContext);
-                getAuthorizationEngine(authentication).resolveAuthorizationInfo(requestInfo, authzInfoListener);
+                engine.resolveAuthorizationInfo(requestInfo, authzInfoListener);
+            }
+        }
+    }
+
+    @Nullable
+    private static AuthorizationContext extractAuthorizationContext(ThreadContext threadContext, String childAction) {
+        final String parentAction = threadContext.getTransient(ORIGINATING_ACTION_KEY);
```

Comment:
Hmmm, "origin" and "parent" actions start to have different meanings as soon as grand children actions become a thing. Out of excessive caution, I would avoid mixing them in this case.
If it is not much of a code change, can we try to preserve the parent action somewhere, maybe inside the `IndexAuthorizationResult`. But maybe there's a better solution that I can't think of atm.

Reply:
`origin` vs `parent` is a good point. This PR's intention is to skip authZ for child and grandchild based on the originating action. So I think it's just a matter of renaming relevant variables and methods to avoid using the word `parent`?

Reply:
I've renamed most of these to refer to the `originating` context.
It's tricky because I think we usually want the originating action rather than the parent but we only have access to the parent's IAC.
But this is the best we can do in the time we have, and we need to do something.

Reply:
Thanks Tim. I now realise that there is a mismatch between the originating action and the IAC which is technically from the parent. I think we are ok for now. It's probably worth to have a follow-up PR that, as Albert suggested, stores the parent action as well. This can potentially apply the "skipping optimisation" to requests like indexing (note the originating action name does not qualify as parent for child actions), e.g.: `indices:data/write/index` -> `indices:data/write/bulk` -> `indics:data/write/bulk[s]` -> `indics:data/write/bulk[s][p]`.